src/tools/chdman.cpp

```diff
@@ -1016,6 +1016,72 @@ static void guess_chs(std::string *filename, uint64_t filesize, int sectorsize,
 }
 
 
+//-------------------------------------------------
+//  reopen_chd_with_parent - find a parent chd in the
+//  directory specified
+//-------------------------------------------------
+
+static chd_error reopen_chd_with_parents(chd_file &input, const char* chd_path, const std::string parents_dirpath, bool writeable = false)
+{
+	if (!input.opened())
+  {
+		input.open(chd_path, writeable);
+	}
+	chd_file *parent = new chd_file;
+	chd_file *child = NULL;
```

Comment:
Please use smart pointers to manage things like this.  It seems the `chd_file` instance used for probing in the first loop is never deleted.  Also `child` can be moved to the scope where it’s used.

Reply:
Changing to smart pointers. About the child, its scope actually has to be the same as parent's, as the last loop will keep changing the parent's reference to the child's, until we are back to the original parent. We actually need to keep a handle to all parents in order for a read to be complete (child read -> parent read -> grandparent read...). Since I limited the chain to 8 for now, I assume it's safe to keep an array of 8 chd_files for this. Any objection?

Reply:
You'll need a way to manage your object lifecycles. Just leaking stuff is not acceptable.

Reply:
The cleanest way would be changing chd_file's m_parent to be a smart pointer instead, but I don't know mame enough to tell if this is a safe change. I believe it is, since it's a private member. If you can give me the green light, I'll change it to shared pointer, which I believe will survive on function exit.

Reply:
That’s not a safe way to do it.  Think more creatively.  Why not return a `std::list` of `chd_file` or something so the calling function can hold onto the parents?

Reply:
Because changing the returning objects would need to trickle down all the way to all the functions that use `parse_input/output_chd_parameters`, which is basically the entire chdman.cpp, like `do_create/extract/add_metadata/etc_*` functions. If that is more desirable, I'll change the entire program.

Reply:
Well leaking objects is not an acceptable way to do it.